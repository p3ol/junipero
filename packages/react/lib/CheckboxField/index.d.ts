import { MutableRefObject, ReactNode, ComponentPropsWithRef } from 'react';
export declare type CheckboxFieldRef = {
    checked: boolean;
    isJunipero: boolean;
    innerRef: MutableRefObject<any>;
    inputRef: MutableRefObject<any>;
};
declare interface CheckboxFieldProps extends ComponentPropsWithRef<any> {
    checked?: boolean;
    children?: ReactNode | JSX.Element;
    className?: string;
    disabled?: boolean;
    id?: string;
    name?: string;
    required?: boolean;
    valid?: boolean;
    value?: any;
    onChange?(changeEvent: {
        value: any;
        checked: boolean;
    }): void;
    onValidate?(value: any, { dirty, required }: {
        dirty?: boolean;
        required?: boolean;
    }): boolean;
    ref?: MutableRefObject<CheckboxFieldRef | undefined>;
}
declare const CheckboxField: import("react").ForwardRefExoticComponent<Omit<CheckboxFieldProps, "ref"> & import("react").RefAttributes<CheckboxFieldRef>>;
export default CheckboxField;
