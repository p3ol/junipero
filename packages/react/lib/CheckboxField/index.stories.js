import { useState } from 'react';
import { action } from '@storybook/addon-actions';

import CheckboxField from '.';
import Abstract from '../Abstract';
import FieldControl from '../FieldControl';
import Label from '../Label';

export default { title: 'react/Checkbox' };

export const basic = () => (
  <CheckboxField onChange={action('change')}>
    Check this
  </CheckboxField>
);

export const disabled = () => (
  <CheckboxField disabled={true}>
    Can&apos;t check this
  </CheckboxField>
);

export const alreadyChecked = () => (
  <CheckboxField onChange={action('change')} checked={true}>
    Uncheck this
  </CheckboxField>
);
export const alreadyCheckedAndDisabled = () => (
  <CheckboxField onChange={action('change')} checked={true} disabled={true}>
    Can&apos;t uncheck this
  </CheckboxField>
);

export const withValue = () => (
  <CheckboxField value="agreement" onChange={action('change')}>
    Check this
  </CheckboxField>
);

export const withAllEvents = () => (
  <CheckboxField
    value="agreement"
    onChange={action('change')}
    onFocus={action('focus')}
    onBlur={action('blur')}
  >
    Check this
  </CheckboxField>
);

export const withChildrenInProps = () => (
  <CheckboxField
    value="agreement"
    children={<p>this is given child</p>}
  />
);

export const withRequiredField = () => (
  <CheckboxField
    value="agreement"
    required={true}
    children={<p>this is given child</p>}
  />
);

export const withLabelAndAbstract = () => (
  <FieldControl>
    <Label className="info">CGU</Label>
    <CheckboxField
      value="agreement"
      required={true}
    >
      I have read and accept the CGU
    </CheckboxField>
    <Abstract className="info">
     You can see our CGU here
    </Abstract>
  </FieldControl>
);

export const controlled = () => {
  const [checked, setChecked] = useState(false);

  return (
    <>
      <button onClick={() => setChecked(c => !c)}>Toggle</button>
      <CheckboxField
        checked={checked}
        onChange={field => {
          setChecked(field.checked);
          action('change')(field);
        }}
      >
        Check this
      </CheckboxField>
    </>
  );
};
