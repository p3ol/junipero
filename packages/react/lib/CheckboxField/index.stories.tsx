import { useState } from 'react';
import { action } from 'storybook/actions';

import CheckboxField from '.';
import Abstract from '../Abstract';
import FieldControl from '../FieldControl';
import Label from '../Label';

export default { title: 'react/Checkbox' };

export const Basic = () => (
  <CheckboxField onChange={action('change')}>
    Check this
  </CheckboxField>
);

export const Disabled = () => (
  <CheckboxField disabled={true}>
    Can&apos;t check this
  </CheckboxField>
);

export const AlreadyChecked = () => (
  <CheckboxField onChange={action('change')} checked={true}>
    Uncheck this
  </CheckboxField>
);
export const AlreadyCheckedAndDisabled = () => (
  <CheckboxField onChange={action('change')} checked={true} disabled={true}>
    Can&apos;t uncheck this
  </CheckboxField>
);

export const WithValue = () => (
  <CheckboxField value="agreement" onChange={action('change')}>
    Check this
  </CheckboxField>
);

export const WithAllEvents = () => (
  <CheckboxField
    value="agreement"
    onChange={action('change')}
    onFocus={action('focus')}
    onBlur={action('blur')}
  >
    Check this
  </CheckboxField>
);

export const WithChildrenInProps = () => (
  <CheckboxField
    value="agreement"
    // eslint-disable-next-line react/no-children-prop
    children={<p>this is given child</p>}
  />
);

export const WithRequiredField = () => (
  <CheckboxField
    value="agreement"
    required={true}
  >
    <p>this is given child</p>
  </CheckboxField>
);

export const WithLabelAndAbstract = () => (
  <FieldControl>
    <Label className="info">CGU</Label>
    <CheckboxField
      value="agreement"
      required={true}
    >
      I have read and accept the CGU
    </CheckboxField>
    <Abstract className="info">
      You can see our CGU here
    </Abstract>
  </FieldControl>
);

export const Controlled = () => {
  const [checked, setChecked] = useState(false);

  return (
    <>
      <button onClick={() => setChecked(c => !c)}>Toggle</button>
      <CheckboxField
        checked={checked}
        onChange={field => {
          setChecked(field.checked);
          action('change')(field);
        }}
      >
        Check this
      </CheckboxField>
    </>
  );
};
