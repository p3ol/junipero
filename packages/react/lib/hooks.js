import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
