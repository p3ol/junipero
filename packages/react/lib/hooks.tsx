import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
