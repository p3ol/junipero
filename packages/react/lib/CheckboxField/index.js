import { forwardRef, useImperativeHandle, useReducer, useRef } from 'react';
import { classNames, mockState } from '@junipero/core';
import { useEventListener } from '@junipero/hooks';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const CheckboxField = forwardRef(({
  checked = false,
  valid = true,
  disabled = false,
  required = false,
  children,
  value,
  id,
  globalEventsTarget,
  className,
  onChange,
  onFocus,
  onBlur,
  onValidate = (val, { required }) => val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();

  const { update: updateControl } = useFieldControl();

  const [state, dispatch] = useReducer(mockState, {
    checked,
    valid,
    dirty: false,
    focused: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    checked: state.checked,
    isJunipero: true,
  }));

  const onBlur_ = e => {
    if (!disabled) {
      dispatch({ focused: false });
      onBlur?.(e);
    }
  };

  const onFocus_ = e => {
    if (!disabled) {
      dispatch({ focused: true });
      onFocus?.(e);
    }
  };

  useEventListener('keypress', e => {

    if (!disabled) {
      onKeyPress_(e);
    }
  }, { target: globalEventsTarget });

  const onKeyPress_ = e => {

    if (state.focused && (e.key === 'Enter' || e.key === ' ')) {
      state.checked = !state.checked;
      state.dirty = true;
      const valid = onValidate?.(
        state.checked, { dirty: state.dirty, required }
      );
      dispatch({ checked: state.checked, valid, dirty: state.dirty });
      onChange?.({ value, checked: state.checked });
      updateControl?.({
        dirty: state.dirty,
        valid,
      });
      e.preventDefault?.();

      return false;
    }

    return true;
  };

  const onChange_ = e => {
    if (disabled) {
      return;
    }

    const checked = e?.target?.checked ?? false;
    state.dirty = true;
    const valid = onValidate?.(checked, { dirty: state.dirty, required });
    dispatch({ checked, valid, dirty: state.dirty });
    onChange?.({ value, checked });

    updateControl?.({
      dirty: state.dirty,
      valid,
    });
  };

  return (
    <label
      htmlFor={id}
      ref={innerRef}
      className={classNames(
        'checkbox-field',
        'junipero',
        {
          disabled,
          checked: state.checked,
          invalid: !state.valid,
        },
        className
      )}
      onFocus={onFocus_}
      onBlur={onBlur_}
      tabIndex={disabled ? -1 : 1}
    >
      <div className="check">
        <input
          { ...rest }
          id={id}
          type="checkbox"
          ref={inputRef}
          value={value}
          checked={state.checked}
          onChange={onChange_}
          tabIndex={-1}
        />
        <div className="inner" />
      </div>
      <div className="content">{children}</div>
    </label>
  );
});

CheckboxField.displayName = 'CheckboxField';
CheckboxField.propTypes = {
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  value: PropTypes.any,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  required: PropTypes.bool,
  onValidate: PropTypes.func,
  valid: PropTypes.bool,
  globalEventsTarget: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.object,
  ]),
};

export default CheckboxField;
