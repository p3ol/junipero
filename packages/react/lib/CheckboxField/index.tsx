import {
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
  forwardRef,
  MutableRefObject,
  ReactNode,
  ComponentPropsWithRef,
} from 'react';
import { classNames, mockState } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';
import { Check } from '../icons';
import { ForwardedProps, MockState } from '../utils';

export declare type CheckboxFieldRef = {
  checked: boolean;
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
  inputRef: MutableRefObject<any>;
};

declare interface CheckboxFieldProps extends ComponentPropsWithRef<any> {
  checked?: boolean;
  children?: ReactNode | JSX.Element;
  className?: string;
  disabled?: boolean;
  id?: string;
  name?: string;
  required?: boolean;
  valid?: boolean;
  value?: any;
  onChange?(changeEvent: { value: any; checked: boolean }): void;
  onValidate?(
    value: any,
    { dirty, required }: { dirty?: boolean; required?: boolean }
  ): boolean;
  ref?: MutableRefObject<CheckboxFieldRef | undefined>;
}
const CheckboxField = forwardRef(({
  checked = false,
  valid = true,
  disabled = false,
  required = false,
  children,
  value,
  id,
  name,
  className,
  onChange,
  onValidate = (val, { required }) => val || !required,
  ...rest
}: CheckboxFieldProps, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();

  type CheckboxState = {
    checked: boolean,
    valid: boolean,
    dirty: boolean
  };
  const [state, dispatch] = useReducer<MockState<CheckboxState>>(mockState, {
    checked: checked ?? false,
    valid: valid ?? true,
    dirty: false,
  });

  useEffect(() => {
    const valid = onValidate(checked ?? false, {
      dirty: state.dirty, required,
    });
    dispatch({ checked: checked ?? false, valid });
    updateControl?.({ dirty: state.dirty, valid });
  }, [checked]);

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    checked: state.checked,
    isJunipero: true,
  }));

  const onKeyPress_ = e => {
    if (e.key === 'Enter' || e.key === ' ') {
      e.preventDefault?.();

      state.checked = !state.checked;
      const valid = onValidate?.(state.checked, { dirty: true, required });
      dispatch({ checked: state.checked, valid, dirty: true });
      onChange?.({ value, checked: state.checked });
      updateControl?.({ dirty: true, valid });

      return false;
    }

    return true;
  };

  const onChange_ = e => {
    if (disabled) {
      return;
    }

    const checked = e?.target?.checked ?? false;
    const valid = onValidate(checked, { dirty: true, required });
    dispatch({ checked, valid, dirty: true });
    onChange?.({ value, checked });
    updateControl?.({ dirty: true, valid });
  };

  return (
    <label
      htmlFor={id}
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'checkbox-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        {
          disabled,
          checked: state.checked,
        },
        className
      )}
      onKeyPress={onKeyPress_}
      tabIndex={disabled ? -1 : 1}
    >
      <input
        id={id}
        name={name}
        type="checkbox"
        ref={inputRef}
        value={checked ? value : ''}
        checked={state.checked}
        onChange={onChange_}
        tabIndex={-1}
      />
      <div className="inner">
        <Check />
      </div>
      <div className="content">{children}</div>
    </label>
  );
}) as ForwardedProps<CheckboxFieldProps, CheckboxFieldRef>;

CheckboxField.displayName = 'CheckboxField';
CheckboxField.propTypes = {
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.any,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  onValidate: PropTypes.func,
  valid: PropTypes.bool,
};
export default CheckboxField;
