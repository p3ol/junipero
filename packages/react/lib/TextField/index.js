import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
