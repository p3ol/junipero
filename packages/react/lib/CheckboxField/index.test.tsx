import { useState } from 'react';
import { act, fireEvent, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CheckboxField from '.';

describe('<CheckboxField />', () => {
  it('should render', () => {
    const { unmount, container } = render(<CheckboxField />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should add value if provided', () => {
    const { unmount, container } = render(<CheckboxField value="value" />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should set checked if provided', () => {
    const { unmount, container } = render(<CheckboxField checked={true} />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should toggle check if checkbox is clicked', async () => {
    const user = userEvent.setup();
    const { unmount, container } = render(<CheckboxField />);
    const input = container.querySelector('input');

    await user.click(input);

    expect(container).toMatchSnapshot('checked');

    await user.click(input);

    expect(container).toMatchSnapshot('unchecked');
    unmount();
  });

  it('should trigger onChange event when user click on checkbox', async () => {
    const onChangeMock = jest.fn();
    const user = userEvent.setup();
    const { unmount, container } = render(
      <CheckboxField onChange={onChangeMock} />
    );
    const input = container.querySelector('input');
    await user.click(input);
    await waitFor(() => expect(onChangeMock).toHaveBeenCalled());

    unmount();
  });

  it('should not trigger event and change state if user click but ' +
  'checkbox is disabled', async () => {
    const onChangeMock = jest.fn();
    const user = userEvent.setup();
    const { unmount, container } = render(
      <CheckboxField onChange={onChangeMock} disabled={true} />
    );
    const input = container.querySelector('input');
    await user.click(input);
    await waitFor(() => expect(onChangeMock).not.toHaveBeenCalled());
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should trigger onChange when user hit enter ' +
  'and checkbox is focused', async () => {
    const onChangeMock = jest.fn();
    const user = userEvent.setup();
    const { unmount, container } = render(
      <CheckboxField onChange={onChangeMock} />
    );

    await act(async () => {
      container.querySelector('label').focus();
    });

    await user.keyboard('{Enter}');
    await waitFor(() => expect(onChangeMock).toHaveBeenCalled());

    unmount();
  });

  it('should trigger onChange when user hit space ' +
  'and checkbox is focused', async () => {
    const onChangeMock = jest.fn();
    const user = userEvent.setup();
    const { unmount, container } = render(
      <CheckboxField onChange={onChangeMock} />
    );

    await act(async () => {
      container.querySelector('label').focus();
    });

    await user.keyboard('{ }');
    await waitFor(() => expect(onChangeMock).toHaveBeenCalled());

    unmount();
  });

  it('should toggle active state on click', async () => {
    const { unmount, container } = render(<CheckboxField />);

    fireEvent.mouseDown(container.querySelector('label'));

    expect(container).toMatchSnapshot('active');
    fireEvent.mouseUp(container.querySelector('label'));

    expect(container).toMatchSnapshot('inactive');
    unmount();
  });

  it('should set focused state on focus', async () => {
    const { unmount, container } = render(<CheckboxField />);

    expect(container).toMatchSnapshot('not focused');

    await act(async () => {
      container.querySelector('label').focus();
    });
    expect(container).toMatchSnapshot('focused');
    unmount();
  });

  it('should set invalid state if ' +
  'checkbox is invalid and required', async () => {
    const user = userEvent.setup();
    const { unmount, container } = render(<CheckboxField required={true} />);
    const input = container.querySelector('input');

    await user.click(input);
    expect(container).toMatchSnapshot('valid');
    await user.click(input);

    expect(container).toMatchSnapshot('invalid');
    unmount();
  });

  it('should set invalid state following ' +
  'custom onValidate function if provided', async () => {
    const user = userEvent.setup();

    const customValidateMock = jest.fn().mockImplementation(
      (val, { dirty }) => !val && dirty
    );
    const { unmount, container } = render(
      <CheckboxField onValidate={customValidateMock} required={true} />
    );
    const input = container.querySelector('input');

    await user.click(input);
    expect(container).toMatchSnapshot('invalid');
    expect(customValidateMock).toHaveBeenCalledWith(
      true, { dirty: true, required: true }
    );

    await user.click(input);
    expect(container).toMatchSnapshot('valid');
    unmount();
  });

  it('should allow to render a controller checkbox', async () => {
    const user = userEvent.setup();
    const handler = jest.fn();

    const ControlledField = () => {
      const [checked, setChecked] = useState(false);

      return (
        <>
          <button onClick={() => setChecked(c => !c)}>Toggle</button>
          <CheckboxField
            checked={checked}
            value="test"
            onChange={field => {
              setChecked(field.checked);
              handler(field);
            }}
          >
            Check this
          </CheckboxField>
        </>
      );
    };

    const { unmount, container } = render(<ControlledField />);
    expect(container).toMatchSnapshot('unchecked');

    await user.click(container.querySelector('button'));
    expect(container).toMatchSnapshot('checked');
    expect(handler).not.toHaveBeenCalled();

    unmount();
  });
});
