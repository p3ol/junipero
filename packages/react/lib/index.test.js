import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
