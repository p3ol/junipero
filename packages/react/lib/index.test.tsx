import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
