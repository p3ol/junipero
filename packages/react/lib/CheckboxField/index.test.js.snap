// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxField /> should add value if provided 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value="value"
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should not trigger event and change state if user click but checkbox is disabled 1`] = `
<div>
  <label
    class="checkbox-field junipero disabled"
    tabindex="-1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should render 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should set checked if provided 1`] = `
<div>
  <label
    class="checkbox-field junipero checked"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        checked=""
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should set focused state on focus: focused 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should set focused state on focus: not focused 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should set invalid state following custom onValidate function if provided: invalid 1`] = `
<div>
  <label
    class="checkbox-field junipero checked invalid"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should set invalid state following custom onValidate function if provided: valid 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should set invalid state if checkbox is invalid and required: invalid 1`] = `
<div>
  <label
    class="checkbox-field junipero invalid"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should set invalid state if checkbox is invalid and required: valid 1`] = `
<div>
  <label
    class="checkbox-field junipero checked"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should toggle active state on click: active 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should toggle active state on click: inactive 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should toggle check if checkbox is clicked: checked 1`] = `
<div>
  <label
    class="checkbox-field junipero checked"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;

exports[`<CheckboxField /> should toggle check if checkbox is clicked: unchecked 1`] = `
<div>
  <label
    class="checkbox-field junipero"
    tabindex="1"
  >
    <div
      class="check"
    >
      <input
        tabindex="-1"
        type="checkbox"
        value=""
      />
      <div
        class="inner"
      />
    </div>
    <div
      class="content"
    />
  </label>
</div>
`;
